src/compiler/program.ts

```diff
@@ -4357,6 +4357,15 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
             }
         }
 
+        if (options.noCheck) {
+            if (options.noEmit) {
+                createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_with_option_1, "noCheck", "noEmit");
+            }
+            if (!options.emitDeclarationOnly) {
+                createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_without_specifying_option_1, "noCheck", "emitDeclarationOnly");
+            }

```

Comment:
Do you know how much more work it'll take to remove this restriction?

Reply:
I have some in-progress work to remove it, so we can prooooobably remove the restriction post-beta (seeings as this is still getting reviewed right now, and I don't think we wanna delay the beta by like a week? Unless we do.). Mostly because reviewing the refactors to the checker to make it happen is going to be annoying, for sure. Making `.references` lazily-calculatable requires taking a lot of logic sprinkled throughout the checker and giving it all a single, central entrypoint (a la `collectLinkedAliases` for the declaration-emit-equivalent elision functionality), which is finicky in terms of preserving the _exact same_ behavior (fun fact: we have many transforms/elisions that apparently just don't apply if a parent node has a grammar error), or at least close enough behavior. Meanwhile, refactoring away from `getNodeCheckFlags` to a `hasNodeCheckFlag` is basically the same thing, but for a handful more uniquely calculated features (though most of them are just variants on "does this node have a child which is X", or "is this node an X within a Y"). A lot of the `NodeCheckFlag` feature-detections could actually be moved into the binder now that we have a name resolver available outside the checker, which is something I want to experiment with.

All this to say: It's definitely doable, and will get done and is being done, but the timeline isn't "one day", and "one week" is a maybe at best, given what's gotta get reviewed.

Reply:
Alternatively: Changing the restriction to "must target esnext and use `verbatimModuleSyntax`" also probably just works.

Reply:
For context, here's the diffs for that: https://github.com/weswigham/TypeScript/commit/64dddfbc0d57e2d986bdeefa475916738267a9b9

Very easy to enable at least *some* js emit targets in that regard (and you can see one baseline actually does differ, but it's an error case for `verbatimModuleSyntax` that's a reexport of a type-only import, so it's what you'd expect to be broken, given the error). @dragomirtitian said this was probably all they wanted, since it's all you need to strip away TS syntax. 🤷‍♂️ 

I, personally, don't want such a limitation though, since the limitation is basically just tech debt on our part.

Reply:
I think I'd rather skip JS emit in this PR than to have a more complicated set of flags to enable; so, all good as-is.